### PR TITLE
feat(api): Add 'count' parameter to return total item count in API v4

### LINKED ISSUE #4506: Add count API or parameter
Well, OK, I admit it, I miss the counts that we used to have in the API. Let's add them back, but with some tweaks to make them better. We have two ways we can do this:

1. A count API that just provides the count for a query.

3. A count parameter on the existing APIs that either:
    - adds the count to the results like v3
    - just returns the count instead of the results. 

The main reason we removed the count was performance and to enable deep pagination, but it makes it difficult to see if filters are working. 

Thinking about this a bit, I think the better ergonomics are option 2. Adding a whole new API for this seems really annoying, so a better trick is to add a new GET parameter to all of our database APIs that returns the count instead of the results. That'd prevent people from turning it on for all results like in v3, which caused problems.

So imagine taking a query:

`/api/rest/v3/opinions-cited/?cited_opinion=32239`

That'd return the items that cite 32239.

But if you wanted the count, you'd just do:

`/api/rest/v3/opinions-cited/?cited_opinion=32239&count=on`

Which would return something like:

`{"count": 3302}`

Easy enough? 

And we could guide our v4 users by providing a reference to the count URL in the normal response. So right now v3 returns:

```
{
    "count": 2287,
    "next": "https://www.courtlistener.com/api/rest/v3/opinions-cited/?cited_opinion=105221&page=2",
    "previous": null,
    "results": [
...
    ]
}
```

That's nice. v4 is the same, but without the count key. In v4 we could do:

```
{
    # Add the count here ⬇️ 
    "count": https://www.courtlistener.com/api/rest/v3/opinions-cited/?cited_opinion=105221&count=on,
    "next": "https://www.courtlistener.com/api/rest/v3/opinions-cited/?cited_opinion=105221&page=2",
    "previous": null,
    "results": [
...
    ]
}
```

Thoughts?

### PR DESCRIPTION
**Pull request by elisa-a-v**
Resolves #4506 

This PR introduces a new `count=on` query parameter to our paginated API endpoints. When this parameter is included in a request, the API returns only the total count of items matching the query filters, without including any result data.

If no `count=on` is added to the request, a count key is now included in the response just like in v3, but instead of providing the count immediately, it provides a URL that points to the count endpoint for the current query parameters to guide users on how to obtain the total count if needed.

V4 removed the count due to the overhead of counting huge querysets in each request, which is why we're adding this tweaked version of the count in the response so the users have access to the information if needed without reintroducing the overhead.

Also updated docs to reflect these changes:
image

### CODE REVIEW

**Review by mlissner — APPROVED** [review]
Looks nice! 

One thought that might be helpful is to confirm the number and type of queries via a test. Could you add that? 

The rest looks great to me, if it looks good to other folks.

**Comment by mlissner** [comment]
Process notes:
 - I assigned this to Alberto since he worked in this area most recently. It's probably best if I continue doing this for PRs.
 - I set him as reviewer and removed Eduardo. It's probably best if I continue doing this for PRs.
 - I added it to the current sprint and put it in the To Do column. It's probably best if you do this in the future. :)

Thank you!

File: cl/api/templates/rest-docs-vlatest.html [hunk 1]
```diff
@@ -324,6 +324,30 @@ <h3 id="ordering">Ordering</h3>
...
+
+{"count": 3302}</pre>
+    <p>When <code>count=on</code> is specified:</p>
+    <ul>
+      <li>The API returns only the <code>count</code> key with the total number of matching items.</li>
+      <li>Pagination parameters like <code>cursor</code> are ignored.</li>
+      <li>The response does not include any result data, which can improve performance for large datasets.</li>
+    </ul>
+    <p>In standard paginated responses, a <code>count</code> key is included with the URL to obtain the total count for your query:</p>
+    <pre class="pre-scrollable">curl "{% get_full_host %}{% url "opinion-list" version="v4" %}?cited_opinion=32239
```
Comment (albertisfu) [inline cl/api/templates/rest-docs-vlatest.html:340]: Just a missing quote.
```suggestion
    <pre class="pre-scrollable">curl "{% get_full_host %}{% url "opinion-list" version="v4" %}?cited_opinion=32239"
```

**Review by albertisfu — COMMENTED** [review]
This looks pretty good to me.

I just left a couple of minor comments.

I also agree with Mike's suggestion to add or tweak the current tests to verify the number and type of queries performed during a regular request and a count request. To ensure that no count query is executed during regular requests, and only the count query is executed during count requests. I think you can base on `ApiQueryCountTests`

Thank you!

File: cl/api/pagination.py [hunk 2]
```diff
@@ -88,6 +90,12 @@ def paginate_queryset(self, queryset, request, view=None):
 
         self.version = request.version
         self.request = request
+        self.is_count_request = request.query_params.get("count") == "on"
```
[the remark this answers is not in the document]
Comment (mlissner) [reply cl/api/pagination.py:93]: It's OK, if it's in v3 too, but if it's this easy, sure, it's better not to bring hot new features to v3.

File: cl/api/templates/rest-docs-vlatest.html (hunk 1 above)
Comment (elisa-a-v) [reply cl/api/templates/rest-docs-vlatest.html:340]: oof thanks for catching that!

**Comment by elisa-a-v** [comment]
> Process notes:
>
> - I assigned this to Alberto since he worked in this area most recently. It's probably best if I continue doing this for PRs.
> - I set him as reviewer and removed Eduardo. It's probably best if I continue doing this for PRs.
> - I added it to the current sprint and put it in the To Do column. It's probably best if you do this in the future. :)

That makes sense, so from now on I'll just add you as a reviewer once a PR is ready for review, and I'll add it to the current sprint 🫡

**Comment by elisa-a-v** [comment]
Thank you @albertisfu and @mlissner for the comments! I'm working on the tests and I think the class you mentioned (`ApiQueryCountTests`) seems like the right place to add the new tests, do you agree?

**Comment by albertisfu** [comment]
> I'm working on the tests and I think the class you mentioned (ApiQueryCountTests) seems like the right place to add the new tests, do you agree?

Yeah, I agree. that's a good place to add the new tests!

**Review by albertisfu — COMMENTED** [review]
Thanks @elisa-a-v the changes and the API count tests added look great.

We can merge this one @mlissner

**Pull request merged**